homeassistant/components/fan/xiaomi_miio.py

```diff
@@ -294,70 +602,235 @@ def speed(self):
+
+        await self._try_command(
+            "Turning the learn mode of the miio device off failed.",
+            self._device.set_learn_mode, False)
+
+    async def async_set_volume(self, volume: int = 50):
+        """Set the sound volume."""
+        if self._device_features & FEATURE_SET_VOLUME == 0:
+            return
+
+        await self._try_command(
+            "Setting the sound volume of the miio device failed.",
+            self._device.set_volume, volume)
+
+    async def async_set_extra_features(self, features: int = 1):
+        """Set the extra features."""
+        if self._device_features & FEATURE_SET_EXTRA_FEATURES == 0:
+            return
+
+        await self._try_command(
+            "Setting the extra features of the miio device failed.",
+            self._device.set_extra_features, features)
+
+    async def async_reset_filter(self):
+        """Reset the filter lifetime and usage."""
+        if self._device_features & FEATURE_RESET_FILTER == 0:
+            return
+
+        await self._try_command(
+            "Resetting the filter lifetime of the miio device failed.",
+            self._device.reset_filter)
+
+
+class XiaomiAirHumidifier(XiaomiGenericDevice):
+    """Representation of a Xiaomi Air Humidifier."""
+
+    def __init__(self, name, device, model, unique_id):
+        """Initialize the plug switch."""
+        super().__init__(name, device, model, unique_id)
+
+        if self._model == MODEL_AIRHUMIDIFIER_CA:
+            self._device_features = FEATURE_FLAGS_AIRHUMIDIFIER_CA
+            self._state_attrs.update({attribute: None for attribute in
+                                      AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER_CA})
+        else:
+            self._device_features = FEATURE_FLAGS_AIRHUMIDIFIER
+            self._state_attrs.update({attribute: None for attribute in
+                                      AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER})
+
+    async def async_update(self):
+        """Fetch state from the device."""
+        from miio import DeviceException
+
+        # On state change the device doesn't provide the new state immediately.
+        if self._skip_update:
+            self._skip_update = False
+            return
+
+        try:
+            state = await self.hass.async_add_job(
+                self._device.status)
+            _LOGGER.debug("Got new state: %s", state)
+
+            self._available = True
+            self._state = state.is_on
+
+            if self._model == MODEL_AIRHUMIDIFIER_CA:
+                self._state_attrs.update(
+                    {key: self._extract_value_from_attribute(state, value)
+                     for key, value in
+                     AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER_CA.items()})
+            else:
+                self._state_attrs.update(
+                    {key: self._extract_value_from_attribute(state, value)
+                     for key, value in
+                     AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER.items()})
```

Comment:
This looks like repetition from above (in `__init__`), I think it'd make sense to have a `self._available_attrs` or something initialized in there.

Reply:
Good catch!